CCDB/src/CcdbApi.cxx

```diff
@@ -643,6 +651,126 @@ void* CcdbApi::extractFromLocalFile(std::string const& filename, TClass const* t
+  auto returncode = system("timeout 1s timeout 1s alien-token-info &> /dev/null");
+  return returncode == 0;
+}
+
+bool CcdbApi::initTGrid() const
+{
+  if (!mAlienInstance) {
+    if (mHaveAlienToken) {
+      mAlienInstance = TGrid::Connect("alien");
+    }
+  }
+  return mAlienInstance != nullptr;
+}
+
+void* CcdbApi::downloadAlienContent(std::string const& url, TClass* cl) const
+{
+  if (!initTGrid()) {
+    return nullptr;
+  }
+  auto memfile = TMemFile::Open(url.c_str(), "OPEN");
+  if (memfile) {
+    auto content = extractFromTFile(*memfile, cl);
+    delete memfile;
+    return content;
+  }
+  return nullptr;
+}
+
+void* CcdbApi::interpretAsTMemFileAndExtract(char* contentptr, size_t contentsize, TClass* tcl) const
+{
+  void* result = nullptr;
+  Int_t previousErrorLevel = gErrorIgnoreLevel;
+  gErrorIgnoreLevel = kFatal;
+  TMemFile memFile("name", contentptr, contentsize, "READ");
+  gErrorIgnoreLevel = previousErrorLevel;
+  if (!memFile.IsZombie()) {
+    result = extractFromTFile(memFile, tcl);
+    if (!result) {
+      LOG(ERROR) << o2::utils::concat_string("Couldn't retrieve object corresponding to ", tcl->GetName(), " from TFile");
+    }
+    memFile.Close();
+  }
+  return result;
+}
+
+// navigate sequence of URLs until TFile content is found; object is extracted and returned
+void* CcdbApi::navigateURLsAndRetrieveContent(CURL* curl_handle, std::string const& url, TClass* cl, std::map<string, string>* headers) const
+{
+  // let's see first of all if the url is something specific that curl cannot handle
+  if (url.find("alien:/", 0) != std::string::npos) {
+    return downloadAlienContent(url, cl);
+  }
+  // add other final cases here
+  // example root://
+
+  // otherwise make an HTTP/CURL request
+  // specify URL to get
+  curl_easy_setopt(curl_handle, CURLOPT_URL, url.c_str());
+  // some servers don't like requests that are made without a user-agent
+  // field, so we provide one
+  curl_easy_setopt(curl_handle, CURLOPT_USERAGENT, "libcurl-agent/1.0");
+  // if redirected , we tell libcurl NOT to follow redirection
+  curl_easy_setopt(curl_handle, CURLOPT_FOLLOWLOCATION, 0L);
+  curl_easy_setopt(curl_handle, CURLOPT_HEADERFUNCTION, header_map_callback<decltype(mHeaderData)>);
+  mHeaderData.clear();
+  curl_easy_setopt(curl_handle, CURLOPT_HEADERDATA, (void*)&mHeaderData);
+
+  MemoryStruct chunk{(char*)malloc(1), 0};
+
+  // send all data to this function
+  curl_easy_setopt(curl_handle, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
+  curl_easy_setopt(curl_handle, CURLOPT_WRITEDATA, (void*)&chunk);
+
+  auto res = curl_easy_perform(curl_handle);
+  long response_code = -1;
+  void* content = nullptr;
+  if (res == CURLE_OK && curl_easy_getinfo(curl_handle, CURLINFO_RESPONSE_CODE, &response_code) == CURLE_OK) {
+    if (headers) {
+      for (auto p : mHeaderData) {
+        (*headers)[p.first] = p.second;
+      }
+    }
+    if (response_code == 404) {
```

Comment:
There are many error cases in addition to `404`. By explicitly handling only `2xx` codes as success this can be the default case. I suggest logging the URL and the response code for all these, to help future debugging.